File: app/src/main/java/com/sjn/stamp/db/SongHistory.java
package com.sjn.stamp.db;

import android.content.res.Resources;

import com.sjn.stamp.R;
import com.sjn.stamp.constant.RecordType;
import com.sjn.stamp.utils.TimeHelper;

import java.util.Date;

import io.realm.RealmObject;
import io.realm.annotations.Index;
import io.realm.annotations.PrimaryKey;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;

@Accessors(prefix = "m")
@AllArgsConstructor(suppressConstructorProperties = true)
@NoArgsConstructor
@Getter
@Setter
public class SongHistory extends RealmObject {

    @PrimaryKey
    public long mId;
    private Song mSong;
    public Date mRecordedAt;
    @Index
    public String mRecordType;
    public Device mDevice;
    public int mCount;
    public float mLatitude;
    public float mLongitude;
    public float mAccuracy;
    public float mAltitude;

    public void setValues(Song song, RecordType recordType, Device device, Date date, int count) {
        setSong(song);
        setRecordedAt(date);
        setRecordType(recordType);
        setDevice(device);
        setCount(count);
    }

    private void setRecordType(RecordType recordType) {
        this.mRecordType = recordType.getValue();
    }

    public String toLabel(Resources resources) {
        return resources.getString(R.string.song_history_label, mSong.getTitle(), mSong.getArtist().getName(), TimeHelper.formatYYYYMMDDHHMMSS(mRecordedAt));
    }
}

File: app/src/main/java/com/sjn/stamp/db/SongHistory.kt
package com.sjn.stamp.db

import android.content.res.Resources
import com.sjn.stamp.R
import com.sjn.stamp.constant.RecordType
import com.sjn.stamp.utils.TimeHelper
import io.realm.RealmObject
import io.realm.annotations.Index
import io.realm.annotations.PrimaryKey
import java.util.*

@Suppress("unused")
open class SongHistory(
        @PrimaryKey var id: Long = 0,
        var song: Song = Song(),
        var recordedAt: Date = Date(),
        @Index private var recordType: String = "",
        var device: Device = Device(),
        var count: Int = 0,
        var latitude: Float = 0.toFloat(),
        var longitude: Float = 0.toFloat(),
        var accuracy: Float = 0.toFloat(),
        var altitude: Float = 0.toFloat()
) : RealmObject() {

    fun applyValues(song: Song, recordType: RecordType, device: Device, date: Date, count: Int): SongHistory {
        this.song = song
        this.recordType = recordType.databaseValue
        this.device = device
        this.recordedAt = date
        this.count = count
        return this
    }

    fun toLabel(resources: Resources): String =
            resources.getString(R.string.song_history_label, song.title, song.artist.name, TimeHelper.formatYYYYMMDDHHMMSS(recordedAt))
}
